### PR TITLE
Fix FilterForm reset

### LINKED ISSUE #7432: [v4] Text Filter Gets Refilled After User Empties It
**What you were expecting:**

When a user empties a text filters, it stays empty

**What happened instead:**

The textinput compes back to pre-reset value on blur

**Steps to reproduce:**

1. Open [the v4 CodeSandbox](https://codesandbox.io/s/github/marmelab/react-admin/tree/next/examples/simple)
2. Type a filter in the seach box (e.g. "tempore"
3. Click on the first element in the list
4. Once in the detail view, **click on the back button**
5. In the list view, empty the text filter with the small cross
6. The previous value ('tempure") comes back

https://user-images.githubusercontent.com/99944/160161603-6eac26d9-56bc-4885-8886-834995e86693.mov

**Environment**

* React-admin version: v4.0.0-beta-3 / `next` branch
* Last version that did not exhibit the issue (if applicable): v3
* React version: 17
* Browser: Chrome

### PR DESCRIPTION
**Pull request by djhi**
Fixes #7432

### CODE REVIEW

File: packages/ra-ui-materialui/src/list/filter/FilterForm.tsx [hunk 1]
```diff
@@ -57,7 +59,8 @@ export const FilterForm = (props: FilterFormProps) => {
             if (isFormValid) {
                 if (lodashGet(values, name) === '') {
                     const newValues = cloneDeep(values);
-                    lodashSet(newValues, name, undefined);
+                    lodashUnset(newValues, name);
+                    console.log({ newValues, displayedFilters });
```
Comment (WiXSL) [inline packages/ra-ui-materialui/src/list/filter/FilterForm.tsx:63]:
```suggestion
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — CHANGES_REQUESTED** [review]
This is great!
However, I found last problem.
1- Add "Title" filter in simple project's PostList.
2- Try to remove the filter entirely by clicking in the remove button.

Result: The filter stays there instead of being removed

File: packages/ra-ui-materialui/src/list/filter/FilterForm.tsx [hunk 2]
```diff
@@ -210,3 +213,38 @@ const StyledForm = styled('form', {
...
+
+/**
+ * Because we are using controlled inputs with react-hook-form, we must provide a default value
+ * for each input when resetting the form. (see https://react-hook-form.com/api/useform/reset).
+ * To ensure we don't provide undefined which will result to the current input value being reapplied
+ * and due to the dynamic nature of the filter form, we rebuild the filter form values from its current
+ * values and make sure to pass at least an empty string for each input.
+ */
+export const getFilterFormValues = (
+    values: Record<string, any>,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/filter/FilterForm.tsx:225]: could you name this one `formValues` for clarity?

File: packages/ra-ui-materialui/src/list/filter/FilterForm.tsx [hunk 3]
```diff
@@ -210,3 +213,38 @@ const StyledForm = styled('form', {
...
+) => {
+    const newValues = cloneDeep(values);
+    Object.keys(values).forEach(key => {
+        setInputValue(newValues, key, filterValues);
+    });
+    return { ...filterValues, ...newValues };
+};
+
+const setInputValue = (
+    values: Record<string, any>,
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/list/filter/FilterForm.tsx:236]: not fan of functions that mutate their arguments

File: packages/ra-ui-materialui/src/list/filter/FilterForm.tsx (hunk 1 above)
Comment (djhi) [reply packages/ra-ui-materialui/src/list/filter/FilterForm.tsx:63]: Oups ^^

File: packages/ra-ui-materialui/src/list/filter/FilterForm.tsx [hunk 4]
```diff
@@ -85,8 +82,8 @@ export type FilterFormProps = FilterFormBaseProps & {
 export const FilterFormBase = (props: FilterFormBaseProps) => {
     const { className, filters, ...rest } = props;
     const resource = useResourceContext(props);
-    const form = useFormContext();
     const { displayedFilters = {}, hideFilter } = useListContext(props);
+    console.log({ displayedFilters });
```
Comment (WiXSL) [inline packages/ra-ui-materialui/src/list/filter/FilterForm.tsx:86]:
```suggestion
```

**Review by WiXSL — CHANGES_REQUESTED** [review]

**Review by WiXSL — CHANGES_REQUESTED** [review]
Now the `Commentable` filter is always On in simple project's PostList

**Review by WiXSL — APPROVED** [review]
Now, is ok 👍

**Review by fzaninotto — APPROVED** [review]

**Pull request merged**